### DISCUSSION TITLE
Combobox with sever-side loading/searching

### DISCUSSION THREAD

**Discussion by marcospassos** [ariakit/ariakit#1415] [Questions] [answered]
I couldn't find any examples of combobox with server-side search. An example of incremental loading (load-as-you-scroll) would also help a lot (we have some unbounded lists here, so we can't just fetch the complete list).

Are there any official or non-official examples out there?

**Comment by diegohaz** [comment, marked as answer]
Does this help? https://codesandbox.io/s/ariakit-combobox-infinity-scroll-03bys?file=/src/App.js

**Reply by marcospassos** [reply]
That's exactly this. How about adding it to the official docs?

**Reply by diegohaz** [reply]
That would be nice! I've added a new "Example request" issue template (https://github.com/ariakit/ariakit/issues/new/choose). Feel free to include any specific requirements you have.

**Reply by marcospassos** [reply]
Oh, nice! Tks!

**Reply by marcospassos** [reply]
@diegohaz, I'm back on this. I'm trying to add a loading indicator at the end of the list to provide feedback to the user. [Here's an example](https://codesandbox.io/s/ariakit-combobox-infinity-scroll-forked-wys8x0?file=/src/App.js).

That simplistic approach works when scrolling, but not when navigating using the keyboard, as the last item keeps focused and the loading indicator never enters the visible area:

https://user-images.githubusercontent.com/943036/173606906-5bf705f3-648b-483b-8339-ee63ca199c4b.mov

So, I tried a different approach, wrapping the loading indicator into a composite item and adding it to the listbox (I use the listbox as the scrollable area instead of the popover, which most of the Ariakit's examples do). Although it worked for the focus issue, it led me to a new problem: whenever the loading finishes, the loading indicator is removed from the list, making the composition loses the active item and restart the navigation from the beginning:

https://user-images.githubusercontent.com/943036/173608150-77f63ae6-9379-47c3-816e-21ff62e0c122.mov

Do you have any suggestions on how I can achieve such a result?

**Reply by diegohaz** [reply]
In your first approach, you can try `<SelectItem preventScrollOnKeyDown={false} />`.

Your second approach looks good. Regarding the focus, the easiest way I can think of is to ensure the `key` prop on the removed loading indicator item matches the `key` prop on the first of the new items. This way, React will treat them as the same DOM element instead of removing and creating a new one (which causes the focus loss). But I haven't tested this with screen readers yet, so I'm not sure it plays well (I mean if the newly added item will be properly announced).

> I use the listbox as the scrollable area instead of the popover, which most of the Ariakit's examples do

I don't quite understand this. The listbox and the popover are usually the same component (`ComboboxPopover`). Are you using them separately?

**Reply by marcospassos** [reply]
> I don't quite understand this. The listbox and the popover are usually the same component (ComboboxPopover). Are you using them separately?

I'm using both (`<SelectPopover><ComboboxList/></SelectPopover>`) since it's select + combobox. I noticed that the `<SelectPopover/>` behaves differently from the `<ComboboxPopover/>` in the sense that the latter is always the listbox itself, while the former doesn't become a listbox when `composite={false}`. Can you confirm?

> In your first approach, you can try <SelectItem preventScrollOnKeyDown={false} />.

That seems to have no effect 🙁

> Your second approach looks good. Regarding the focus, the easiest way I can think of is to ensure the key prop on the removed loading indicator item matches the key prop on the first of the new items.

I don't think it will be easier at all due to how our components are decoupled: the listbox is completely unaware of the surroundings so that it can be reused in select and comboboxes, or any other component that requires a list of options.

I'm wondering if we have some way to make the composite state keep the _previous-still-in-the-list_ item whenever an item is removed, so removing the loading indicator would set the active item back to the previous item.

Another alternative would be scrolling the list to the bottom after pressing keydown on the last item (assuming `focusWrap={false}`). However, `preventScrollOnKeyDown={false}` doesn't seem to help much here.

Any clues?

**Reply by diegohaz** [reply]
> Another alternative would be scrolling the list to the bottom after pressing keydown on the last item (assuming `focusWrap={false}`). However, `preventScrollOnKeyDown={false}` doesn't seem to help much here.

This looks like a bug. Created #1505.

**Reply by marcospassos** [reply]
@diegohaz This solution seems to work pretty well, wdy?
https://codesandbox.io/s/ariakit-combobox-infinity-scroll-forked-wys8x0?file=/src/App.js

**Reply by diegohaz** [reply]
That looks good! You should probably check the `PageDown` key as well.

**Reply by marcospassos** [reply]
Well minded!
